rockin/plugins/drilling_machine/drilling_machine_simulation_thread.cpp

```diff
@@ -0,0 +1,144 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Library General Public License for more details.
+ *
+ *  Read the full text in the LICENSE.GPL file in the doc directory.
+ */
+
+#include "drilling_machine_simulation_thread.h"
+
+#include <core/threading/mutex_locker.h>
+#include <boost/lexical_cast.hpp>
+#include <boost/thread/thread.hpp>
+
+/** @class DrillingMachineSimulationThread
+ * Thread to simulate RoCKIn@Work drilling machine.
+ * @author Alexander Moriarty
+ */
+fawkes::Mutex DrillingMachineSimulationThread::drill_machine_mutex_;
+
+/** Constructor. */
+DrillingMachineSimulationThread::DrillingMachineSimulationThread() :
+    Thread("DrillingMachineSimulationThread", Thread::OPMODE_CONTINUOUS),
+    cfg_timer_interval_(40)
+{
+}
+
+void DrillingMachineSimulationThread::init()
+{
+    try
+    {
+        cfg_timer_interval_ = config->get_uint("/llsfrb/clips/timer-interval");
+    } catch (fawkes::Exception &e)
+    {
+        // use default value
+    }
+
+    last_sent_command_timestamp_ = boost::posix_time::microsec_clock::local_time();
+
+    logger->log_info("DrillingMachineSim", "Plugin is in simulation mode.");
+
+    fawkes::MutexLocker lock2(&drill_machine_mutex_);
+
+    last_status_msg_.set_state(DrillingMachineStatus::AT_TOP);
+    last_status_msg_.set_is_device_connected(true);
+    tick_ = ENCODER_MAX_;
+    last_command_msg_.set_command(DrillingMachineCommand::MOVE_UP);
+    logger->log_info("DrillingMachineSim", "simulation initialized to AT_TOP, tick at %d", tick_);
+
+    fawkes::MutexLocker lock(clips_mutex);
+
+    clips->add_function("drilling-machine-move-drill-up", sigc::slot<void>(sigc::mem_fun(*this, &DrillingMachineSimulationThread::clips_move_drill_up)));
+    clips->add_function("drilling-machine-move-drill-down", sigc::slot<void>(sigc::mem_fun(*this, &DrillingMachineSimulationThread::clips_move_drill_down)));
+    clips->add_function("drilling-machine-get-state", sigc::slot<int>(sigc::mem_fun(*this, &DrillingMachineSimulationThread::clips_get_device_state)));
+    clips->add_function("drilling-machine-is-device-connected", sigc::slot<int>(sigc::mem_fun(*this, &DrillingMachineSimulationThread::clips_is_device_connected)));
+
+    if (!clips->build("(deffacts have-feature-drilling-machine (have-feature DrillingMachine))"))
+        logger->log_warn("DrillingMachineSim", "Failed to build deffacts have-feature-drilling-machine");
+}
+
+void DrillingMachineSimulationThread::finalize()
+{
+}
+
+void DrillingMachineSimulationThread::loop()
+{
+    fawkes::MutexLocker lock2(&drill_machine_mutex_);
```

Comment:
This together with the last line in this function leads to a horrible lockup of the complete CFH:
* The _scoped_ lock is acquired here
* In the last line the thread goes to sleep, but does not release the lock. Therefore, all calls that also need to lock the mutex will wait.

My proposal is to manually release the lock before sending the thread to sleep.

Additionally, why is the variable here (and in other places) called lock2?